grant-project/grant-monorepo
# Create overhaul (Pt. 1 - Server side drafts, proposal table fields)

This is an awful, awful PR but it's a necessary step to completing #196, #195, #169, #146, #84, #79, _and_ #31. It also opens the door to non-smart contract chains.

### What This Does

Tears up the floorboards.

* Create drafts of proposals. Publishing just updates the state of the proposal from `DRAFT` to `LIVE`.
* Removes the ability to create users via proposal creation. This also means no adding team members (temporarily.)
* Adds all contract & currently un-recorded information to the `proposal` table (brief, target, payout address etc.)
* Adds additional input validation server side (Still not as much as I'd like though.)
* Removes the `CreateFormState` type altogether, create form ingests `ProposalDraft` now which is more or less the same as `Proposal`, minus some stuff.
* All fields match the column names. No more `details`, `body` aliases for content etc.

This one's a pretty big PR, so I totally understand if you break your reviews out into multiple parts.

### What's Coming Next

* Add users to your team by email / account address.
    * Merging `User` and `TeamMember` types as much as possible.
* Edit live proposals.
* Handling for when a contract deploy is successful, but the publishing is not.

### Steps to Test

Unfortunately due to SQLLite making the migration to these new fields a pain, I've opted to blow away the migrations for this one. The easiest way to test this without blowing out your db would be to temporarily change your environment variable db path while testing.

* Create a proposal
* Go back to `/create`, confirm you're shown the proposal draft you just made.
* Fill out the proposal, refresh, confirm the data is still there.
* Go to `/create`, confirm it also shows the entered data.
* Publish the proposal, confirm that works.
* Try to edit the proposal by going to `/proposals/{id}/edit`, confirm this does not work.
* Go back to `/create`, confirm your proposal is no longer there.
* Create a new proposal, go to `/create`, and delete it. Confirm it was deleted.

Ideally try to mess around with the forms as much as possible while you're doing this.


## Screenshots

### Drafts!

<img width="793" alt="screen shot 2018-11-15 at 10 55 52 am" src="https://user-images.githubusercontent.com/649992/48564829-2ca4cf80-e8c5-11e8-8781-50712db67a2e.png">
